### PR TITLE
feat(package): list packages

### LINKED ISSUE #7: implement command to list packages
```bash
xst packages list [options]
```

Aliases: 

- `ls`

Should return tabulated list of **package-abbreviations**.

Example Output:

```
package-service
dashboard
monex
html-templating
eXide
```

Possible options:

- `--versions`, `-V` :  list installed version
- `--libraries`, `-l` or `-L`: list only installed libraries
- `--applicationts`, `-a` or `-A`: list only installed applications
- `--filter`, `-f`: list only libraries and/or applications that match a given pattern
- `--extended`, `-e`: add more information to the result (date of installation, namespace URI, declared dependencies, ...)
- `--json`, `-j`: return result as json (in order to pipe it to `jq` for example or store it in a file)
- `--full-name`: return namespace URIs instead of abbreviated names 

refs https://github.com/eXist-db/node-exist/issues/223

### PR DESCRIPTION
**Pull request by line-o**
closes #7

```bash
xst package list [options]

List installed packages

Options:
      --version               Show version number                                          [boolean]
      --config                configuration to use
      --help                  Show help                                                    [boolean]
  -G, --color                 Color the output                            [boolean] [default: false]
  -V, --versions              Display installed version                   [boolean] [default: false]
  -e, --extended              Display more information for each item      [boolean] [default: false]
  -t, --typesort              Sort by type                                [boolean] [default: false]
  -T, --timesort              Sort by installation date                   [boolean] [default: false]
  -r, --reverse               Reverse the order of the sort                                [boolean]
  -a, --app, --applications   Only list installed application packages                     [boolean]
  -l, --lib, --libraries      Only list installed library packages                         [boolean]
      --date                  Show installation date                       [choices: "short", "iso"]
  -D, --deps, --dependencies  Show package dependencies                   [boolean] [default: false]
      --full-name             Use full name for packages and dependencies [boolean] [default: false]
      --raw                   Return raw JSON data from query             [boolean] [default: false]
```

Implements all options in #7 and then some. Except the ability to filter the list using a globbing expression.
This should rather be done as a separate command `xst package info <package-abbrev>`.

### CODE REVIEW

**Comment by line-o** [comment]
example output of the templating library with `--extended`

```
templating
Title: eXist-db HTML Templating Library
URI: http://exist-db.org/html-templating
Author: eXist-db Project
Description: This package contains the HTML templating library for eXist, which was previously part of the now deprecated 'shared-resources'
Website: https://github.com/eXist-db/templating
Version: 1.1.0
Installed: 2023-01-04T17:05:21.129+01:00
Processor: existdb >=5.2.0
Type: library
License: LGPL-2.1
Components:
  xquery:
    http://exist-db.org/xquery/html-templating
    http://exist-db.org/xquery/html-templating/lib
```

**Comment by line-o** [comment]
Example output of `xst package list --app --date --versions --dependencies`

```
dashboard           2.0.9  Jan  4 17:05
├── http://www.functx.com >=1.0
└── http://exist-db.org/apps/existdb-packageservice >=1.3.8
eXide               3.4.0  Jan  4 17:05
└── (no dependency)
exist-documentation 5.4.0  Jan  4 17:05
└── http://exist-db.org/html-templating >=1.0.2
fundocs             1.2.0  Jan  4 17:05
└── http://exist-db.org/html-templating >=1.0.2
monex               3.0.5  Jan  4 17:05
└── http://exist-db.org/html-templating >=1.0.2
packageservice      1.3.14 Jan  4 17:05
├── http://www.functx.com *
└── http://exist-db.org/xquery/semver-xq *
```

File: commands/package/list.js [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+  }
+  if (versionedItem.min && versionedItem.max) {
+    return versionedItem.min + '-' + versionedItem.max
+  }
+  if (versionedItem.min) {
+    return '>=' + versionedItem.min
+  }
+  if (versionedItem.max) {
+    return '<=' + versionedItem.max
+  }
```
Comment (joewiz) [inline commands/package/list.js]: Perhaps I'm reading this code wrong, but it appears that you're assuming that (a) only `@semver` can be an template and (b) `@semver-min` and `@semver-max` cannot be templates. But based on the EXPath Package spec's [section on templates](http://expath.org/spec/pkg#pkgdep), these attributes could all be templates. For a real life example of a `@semver-max` that could be a template, see https://github.com/eXist-db/public-repo/issues/72. Also, in this case, the rendering of `semver-max="5"` is `<6.0.0`, not `<=5.0.0`. See https://github.com/eXist-db/semver.xq/blob/master/src/main/xquery/semver.xqm#L208-L247.

Comment (line-o) [reply commands/package/list.js]: Very interesting find! Thanks for having a thorough look at it.

Comment (line-o) [reply commands/package/list.js]: I added unit tests for version checks and formatting to ensure it works and found additional problems.

Comment (line-o) [reply commands/package/list.js]: They are all fixed now with 5401b6c40634a90c5c7c0fef9cc0f6469fcb9e35

**Comment by line-o** [comment]
Example `--extended` output for **functx**

```
functx
Title: FunctX library
URI: http://www.functx.com
Author: Priscilla Walmsley
Description: The FunctX XQuery/XSLT 2.0 function library is a set of reusable functions for XQuery 1.0 and XSLT 2.0. The project itself is an open source application and framework for developers to create function libraries.
Website: http://www.functx.com/
Version: 1.0.1
Installed: 2023-01-06T21:41:14.261+01:00
Processor: any
Type: library
License: GNU-LGPL
Components:
  xslt: http://www.functx.com/functx.xsl
  xquery: http://www.functx.com
```

**Comment by line-o** [comment]
`--raw` can be used to pipe the json result to `jq`

```bash
xst package list --raw | jq '.packages[0]' -M
```

```
{
  "date": "2023-01-04T17:05:22.989+01:00",
  "processor": {
    "uri": "http://exist-db.org",
    "min": "5.3.0-SNAPSHOT",
    "max": null,
    "template": null,
    "exact": []
  },
  "components": {
    "xslt": [],
    "xquery": [],
    "xproc": [],
    "xsd": [],
    "rng": [],
    "schematron": [],
    "nvdl": [],
    "resource": [],
    "jar": []
  },
  "website": "https://www.github.com/exist-db/documentation",
  "abbrev": "fundocs",
  "target": "fundocs",
  "description": null,
  "type": "application",
  "uri": "http://exist-db.org/apps/fundocs",
  "authors": [
    "eXist-db"
  ],
  "version": "1.2.0",
  "title": "XQuery Function Documentation",
  "dependencies": [
    {
      "uri": "http://exist-db.org/html-templating",
      "min": "1.0.2",
      "max": null,
      "template": null,
      "exact": []
    }
  ],
  "license": "GNU Lesser General Public License, version 2.1"
}
```

File: commands/package/list.js (hunk 1 above)
Comment (line-o) [reply commands/package/list.js]: `@semver-max="5"` translates to `<=5` now and the tests show that `5.99.999` satisfies.
The new function `formatVersionFill` will return `<=5.x.x` but this is not yet used anywhere. This could be made an option in `package list` in the future.

Comment (line-o) [reply commands/package/list.js]: While `<6.0.0` could be used and does transport the meaning, displaying this in the list could create confusion as the string "6" is not in the dependency declarations.

Comment (joewiz) [reply commands/package/list.js]: @line-o  `@semver-max="5"` should translate to `<=6`, not `<=5`.

Comment (joewiz) [reply commands/package/list.js]: (The one oddity that the EXPath Package spec authors may not have envisioned - since it predated SemVer 2.0 and maybe even 1.0? - is that `6.0.0-SNAPSHOT` satisfies `@semver-max=5`.)

Comment (joewiz) [reply commands/package/list.js]: (EXPath Package "SemVer templates" are not particularly expressive. npm's SemVer ranges are much more. But they're also [not in the SemVer spec](https://github.com/semver/semver/issues/205). It would be nice to have a spec to reference when considering a successor to EXPath Package SemVer templates for an EXPath Package 2.0... In the absence of this, the EXPath Package spec and SemVer 2.0 is all we've really got.)

**Comment by joewiz** [comment]
In this raw output, "template" is again a bit misleading. Best to call the properties what they are in the spec: `semver-min`, `semver-max`, `semver`, `versions`, and `name`.

**Comment by joewiz** [comment]
`URI:` also doesn't conform to the spec. The property is `@name`, not `@uri`. In public-repo, we use "Name (URI)" to distinguish the `@name` property from the package's `<title>`. I'd suggest either `Name:` or `Name (URI):`. (The package spec says the property is an IRI, but in practice I've only ever seen URIs... not to say I can tell you what the difference is without googling.)

File: commands/package/list.js (hunk 1 above)
Comment (line-o) [reply commands/package/list.js]: Whereas `6.0.0-SNAPSHOT` does __not__  satisfy `<=5`. The intended outcome. @joewiz I would really like to stick to npm SemVer ranges in the output as this is - in my mind - the best we have.

Comment (line-o) [reply commands/package/list.js]: Can you point me to the resource in either EXPath Package 2.0 or SemVer 2.0 from where I can see that  `semver-max="5"` translates to `<6`?
If this is implemented in that way in the public repo, I would consider it an issue.

**Comment by line-o** [comment]
I will rename `uri` to `name`. I think that is sensible to do as it is in fact a free-text property.

**Comment by line-o** [comment]
> In this raw output, "template" is again a bit misleading. Best to call the properties what they are in the spec: `semver-min`, `semver-max`, `semver`, `versions`, and `name`.

I asked about a better name for template above :)
Property names in JSON with hyphens are possible, but I would want to avoid them as it makes it less comfortable to work with in JS. It is also somewhat uncommon.
We could either
- camelCase them (`semverMin`, `semverMax`, `semver`, `versions`)
- shorten them (`min`, `max`, `range`, `versions`)

I do see the value of staying close to the input XML, but it is already aggregated. `components.xquery` for example combines the information from expath-pkg.xml and exist.xml to show all registered module names in one place.

**Comment by line-o** [comment]
The property names of dependencies now reflect the attribute names of declared dependencies @joewiz (see updated --raw output).

**Comment by line-o** [comment]
@joewiz all suggested changes are now implemented. The package itself still had a `uri` property which is now renamed to `name`.
Also the command line flag `--full-uri` was renamed to `--full-name`

File: commands/package/list.js (hunk 1 above)
Comment (line-o) [reply commands/package/list.js]: > (The one oddity that the EXPath Package spec authors may not have envisioned - since it predated SemVer 2.0 and maybe even 1.0? - is that 6.0.0-SNAPSHOT satisfies @semver-max=5.)

@joewiz Can you please explain how you got to this conclusion?

Comment (duncdrum) [reply commands/package/list.js]: The expath spec doesn’t mention pre-release labels. So my reading is that a template `5` means anything with a 5 major version. If we want to include `5.0.0-Snapshot` or `6.0.0-Snapshot` in that is up to us. In light of the communities development practices I m against including the latter. As per SemVer definition, it includes breaking aka incompatible changes. I m not happy with including the former but I can live with it. 

TLDR; +1 for `<=`
